### PR TITLE
Move WakeIOThread() functionality to WatchableEventManager

### PR DESCRIPTION
**Pull request by kpschoedel**
#### Problem

System::Layer contains `WakeIOThread()` and `mWakeEvent` which are
really implementation details of `WatchableEventManager`.

#### Change overview

- Replace `System::Layer::WakeIOThread()` with
  `WatchableEventManager::Signal()`, because not every implementation
  will necessarily implement the nudge by waking a thread.

- Split `SystemSockets.h` into `WatchableEventManager.h`,
  `WatchableSocket.h`, and `WakeEvent.h`; likewise split the
  various implementation files.

- Make the `System::WakeEvent` public API independent of the
  file-descriptor implementation.

For the present, `mHandleSelectThread` remains in `System::Layer`
until timers are integrated and the transitional `HandleTimeout()`
is removed.

#### Testing

Sanity check with chip-tool. CI should verify that functionality does
not regress.

### CODE REVIEW

**Comment by kpschoedel** [comment]
Reviewers — github's presentation of this change obscures what's going on, since you can't tell which parts of the split files are actually new. Give me a moment to make gists with readable diffs.

File: src/system/WatchableEventManagerLibevent.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file declares an implementation of WatchableEvents using libevent.
```
Comment (andy31415) [inline src/system/WatchableEventManagerLibevent.h:20]: nit: do we need this `@file` description given that the file name is already "Watchable Event Manager Libeevent" ?

**Review by andy31415 — APPROVED** [review]

**Comment by kpschoedel** [comment]
- `SystemSockets.h` is split into
  - [`WatchableSocket.h`](https://gist.github.com/kpschoedel/895f3e010c5bce147804fc87a160215f)
  - [`WatchableEventManager.h`](https://gist.github.com/kpschoedel/ecc397e3417f98f783eab73c475ebd25)
  - [`WakeEvent.h`](https://gist.github.com/kpschoedel/234612fb3cc3a7ec4ff0f16fea731525)


- `WatchableSocketLibevent.h` spawns [`WatchableEventManagerLibevent.h`](https://gist.github.com/kpschoedel/451cf083e707c2a5405267dca5384f2a)
- `WatchableSocketLibevent.cpp` spawns [`WatchableEventManagerLibevent.cpp`](https://gist.github.com/kpschoedel/7e7b22d4ea056d9f37e46bae88112d53)

- `WatchableSocketSelect.h` spawns [`WatchableEventManagerSelect.h`](https://gist.github.com/kpschoedel/9e64b03251925ab5b3df7bc61f0a0f78)
- `WatchableSocketSelect.cpp` spawns [`WatchableEventManagerSelect.cpp`](https://gist.github.com/kpschoedel/e97522d8b251a7a30d1735ecd765d56d)

File: src/system/WatchableEventManagerLibevent.h (hunk 1 above)
Comment (kpschoedel) [reply src/system/WatchableEventManagerLibevent.h:20]: We don't, but this particular line is not actually touched in the PR, it's just carried over from the parent file this was split from ([`WatchableSocketLibevent.h` → `WatchableEventManagerLibevent.h`](https://gist.github.com/kpschoedel/451cf083e707c2a5405267dca5384f2a))

File: src/system/WakeEvent.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/*
```
Comment (kpschoedel) [inline src/system/WakeEvent.h:1]: Derived from `SystemSockets.h`; see https://gist.github.com/kpschoedel/234612fb3cc3a7ec4ff0f16fea731525 for diff

File: src/system/WatchableEventManager.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
+/*
```
Comment (kpschoedel) [inline src/system/WatchableEventManager.h:1]: Derived from `SystemSockets.h`; see https://gist.github.com/kpschoedel/ecc397e3417f98f783eab73c475ebd25 for diff

File: src/system/WatchableEventManagerLibevent.cpp [hunk 4]
```diff
@@ -0,0 +1,196 @@
+/*
```
Comment (kpschoedel) [inline src/system/WatchableEventManagerLibevent.cpp:1]: Derived from `WatchableSocketLibevent.cpp`; see https://gist.github.com/kpschoedel/7e7b22d4ea056d9f37e46bae88112d53 for diff

File: src/system/WatchableEventManagerLibevent.h [hunk 5]
```diff
@@ -0,0 +1,73 @@
+/*
```
Comment (kpschoedel) [inline src/system/WatchableEventManagerLibevent.h:1]: Derived from `WatchableSocketLibevent.h`; see https://gist.github.com/kpschoedel/451cf083e707c2a5405267dca5384f2a for diff

File: src/system/WatchableEventManagerSelect.cpp [hunk 6]
```diff
@@ -0,0 +1,233 @@
+/*
```
Comment (kpschoedel) [inline src/system/WatchableEventManagerSelect.cpp:1]: Derived from `WatchableSocketSelect.cpp`; see https://gist.github.com/kpschoedel/e97522d8b251a7a30d1735ecd765d56d for diff

File: src/system/WatchableEventManagerSelect.h [hunk 7]
```diff
@@ -0,0 +1,92 @@
+/*
```
Comment (kpschoedel) [inline src/system/WatchableEventManagerSelect.h:1]: Derived from `WatchableSocketSelect.h`; see https://gist.github.com/kpschoedel/9e64b03251925ab5b3df7bc61f0a0f78 for diff

**Review by bzbarsky-apple — APPROVED** [review]
Thank you for the links to actually-useful diffs!

**Review by jmartinez-silabs — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Pull request merged**